File: blue-sdk-android/blue-sdk/src/main/kotlin/com/blue/sdk/manager/DeviceManager.kt
// DeviceManager.kt
// BlueSDK - 设备信息查询与时间同步（FR12~FR14）

package com.blue.sdk.manager

import com.blue.sdk.error.BlueError
import com.blue.sdk.internal.CommandQueue
import com.blue.sdk.model.DeviceInfo
import com.blue.sdk.transport.CommandCode
import com.blue.sdk.transport.FrameBuilder
import java.util.Calendar
import java.util.TimeZone

internal class DeviceManager(private val commandQueue: CommandQueue) {

    /**
     * 查询设备基础信息（FR12）
     * 应答格式：[MAC 6字节][...][固件版本 ASCII]
     * 示例：6F 74 36 74 74 64 35 6A 31 2E 30 2E 30
     *       MAC=6F:74:36:74:74:64, 固件版本="1.0.0"
     */
    fun queryDeviceInfo(completion: (Result<DeviceInfo>) -> Unit) {
        val frame = FrameBuilder.build(CommandCode.QUERY_DEVICE_INFO)
        commandQueue.enqueue(CommandCode.QUERY_DEVICE_INFO, frame) { result ->
            result.fold(
                onSuccess = { response ->
                    val data = response.data
                    if (data.size < 6) {
                        completion(Result.failure(BlueError.InvalidParameter))
                        return@fold
                    }
                    // 前 6 字节为设备 MAC 地址
                    val macBytes = data.copyOfRange(0, 6)
                    val macAddress = macBytes.joinToString(":") { "%02X".format(it) }
                    // 从第7字节开始查找固件版本（匹配 "x.x.x" 格式）
                    val remaining = data.copyOfRange(6, data.size)
                    val asciiStr = String(remaining, Charsets.US_ASCII)
                    val version = DeviceInfo.parseFirmwareVersion(asciiStr)
                    completion(Result.success(DeviceInfo(
                        firmwareVersion = version,
                        macAddress = macAddress
                    )))
                },
                onFailure = { completion(Result.failure(it as BlueError)) }
            )
        }
    }

    /**
     * 向设备下发当前系统时间（FR14）
     * 使用 sendDirect 直接发送，不经过指令队列排队等待应答
     */
    fun syncTime(timeMs: Long = System.currentTimeMillis(), completion: (Result<Unit>) -> Unit) {
        val data = buildTimeSyncData(timeMs)
        val frame = FrameBuilder.build(CommandCode.TIME_SYNC, data)
        commandQueue.sendDirect(frame)
        completion(Result.success(Unit))
    }

    /**
     * 构建时间同步数据字节
     * 格式（11字节）：[0x00][0x00][年偏移(从2018)][月][日][时][分][秒][星期(1=周一~7=周日)][时区高][时区低]
     */
    private fun buildTimeSyncData(timeMs: Long): ByteArray {
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = timeMs

        val year = calendar.get(Calendar.YEAR)
        val yearOffset = (year - 2018).coerceIn(0, 255).toByte()
        val month = (calendar.get(Calendar.MONTH) + 1).toByte()
        val day = calendar.get(Calendar.DAY_OF_MONTH).toByte()
        val hour = calendar.get(Calendar.HOUR_OF_DAY).toByte()
        val minute = calendar.get(Calendar.MINUTE).toByte()
        val second = calendar.get(Calendar.SECOND).toByte()
        // 星期：Calendar.SUNDAY=1...SATURDAY=7 → 协议 1=周一...7=周日
        val calWeekday = calendar.get(Calendar.DAY_OF_WEEK)
        val weekday = (if (calWeekday == Calendar.SUNDAY) 7 else calWeekday - 1).toByte()

        // 时区偏移（分钟）
        val tzOffset = TimeZone.getDefault().getOffset(timeMs) / 60000
        val tzHigh = ((tzOffset shr 8) and 0xFF).toByte()
        val tzLow = (tzOffset and 0xFF).toByte()

        return byteArrayOf(0x00, 0x00, yearOffset, month, day, hour, minute, second, weekday, tzHigh, tzLow)
    }
}

File: blue-sdk-ios/BlueSDK/BlueSDK/Classes/Manager/DeviceManager.swift
// DeviceManager.swift
// BlueSDK - LX-PD02 智能药盒蓝牙通信 SDK
//
// 设备信息查询与时间同步管理器（FR12~FR14）

import Foundation

/// 设备管理器
final class DeviceManager {

    private let commandQueue: CommandQueue
    private let logger = BlueLogger.shared

    init(commandQueue: CommandQueue) {
        self.commandQueue = commandQueue
    }

    // MARK: - 设备信息查询（FR12）

    /// 查询设备基础信息
    /// 应答格式：[MAC 6字节][...][固件版本 ASCII]
    /// 示例：6F 74 36 74 74 64 35 6A 31 2E 30 2E 30
    ///       MAC=6F:74:36:74:74:64, 固件版本="1.0.0"
    /// - Parameter completion: 结果回调，成功返回 DeviceInfo
    func queryDeviceInfo(completion: @escaping (Result<DeviceInfo, BlueError>) -> Void) {
        let frame = FrameBuilder.build(cmd: CommandCode.queryDeviceInfo)
        commandQueue.enqueue(cmd: CommandCode.queryDeviceInfo, frame: frame) { result in
            switch result {
            case .success(let response):
                let data = response.data
                guard data.count >= 6 else {
                    completion(.failure(.invalidParameter))
                    return
                }
                // 前 6 字节为设备 MAC 地址
                let macAddress = Array(data.prefix(6))
                // 从第7字节开始查找固件版本（匹配 "x.x.x" 格式）
                let remaining = Array(data.suffix(from: 6))
                let asciiStr = String(bytes: remaining, encoding: .ascii) ?? ""
                let version: String
                if let range = asciiStr.range(of: "[0-9]+\\.[0-9]+\\.[0-9]+", options: .regularExpression) {
                    version = String(asciiStr[range])
                } else {
                    version = "Unknown"
                }
                let info = DeviceInfo(firmwareVersion: version, macAddress: macAddress)
                completion(.success(info))
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }

    // MARK: - 时间同步（FR13、FR14）

    /// 向设备下发当前系统时间
    /// 注意：时间同步为单向下发，设备不返回应答。直接发送，不入 CommandQueue 等待。
    /// - Parameters:
    ///   - date: 要同步的时间，默认当前系统时间
    ///   - completion: 结果回调（发送成功即视为成功）
    func syncTime(date: Date = Date(), completion: @escaping (Result<Void, BlueError>) -> Void) {
        let data = buildTimeSyncData(from: date)
        let frame = FrameBuilder.build(cmd: CommandCode.timeSync, data: data)
        // 直接发送，不等待应答（协议约定设备不对时间同步作应答）
        commandQueue.sendDirect(frame: frame)
        completion(.success(()))
    }

    // MARK: - 私有方法

    /// 构建时间同步数据字节
    /// 格式（11字节）：[0x00][0x00][年偏移(从2018)][月][日][时][分][秒][星期(1=周一~7=周日)][时区高][时区低]
    /// 参考协议文档示例：55 AA 00 E1 00 0B 00 00 01 0C 1E 0F 34 1F 01 03 20 9C
    /// 解析：年=2018+1=2019, 月=12, 日=30, 时=15, 分=52, 秒=31, 星期一
    private func buildTimeSyncData(from date: Date) -> [UInt8] {
        let calendar = Calendar.current
        let components = calendar.dateComponents(
            [.year, .month, .day, .hour, .minute, .second, .weekday],
            from: date
        )

        let year = components.year ?? 2024
        let yearOffset = UInt8(min(max(year - 2018, 0), 255))
        let month    = UInt8(components.month ?? 1)
        let day      = UInt8(components.day ?? 1)
        let hour     = UInt8(components.hour ?? 0)
        let minute   = UInt8(components.minute ?? 0)
        let second   = UInt8(components.second ?? 0)
        // 星期：Calendar 中 1=周日...7=周六 → 协议中 1=周一...7=周日
        let calWeekday = components.weekday ?? 1
        let weekday = UInt8(calWeekday == 1 ? 7 : calWeekday - 1)

        // 时区偏移（秒），转换为分钟
        let timeZoneOffset = TimeZone.current.secondsFromGMT() / 60
        let tzHigh = UInt8((timeZoneOffset >> 8) & 0xFF)
        let tzLow  = UInt8(timeZoneOffset & 0xFF)

        return [0x00, 0x00, yearOffset, month, day, hour, minute, second, weekday, tzHigh, tzLow]
    }
}
